addons/tls13-middlebox/bootstrap.js

```diff
@@ -0,0 +1,241 @@
+"use strict";
+
+let {classes: Cc, interfaces: Ci, utils: Cu, results: Cr} = Components;
+
+Cu.import("resource://gre/modules/Preferences.jsm");
+Cu.import("resource://gre/modules/TelemetryController.jsm");
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+
+const VERSION_MAX_PREF = "security.tls.version.max";
+const FALLBACK_LIMIT_PREF = "security.tls.version.fallback-limit";
+
+let readwrite_prefs = new Preferences({defaultBranch: true});
+
+// all combination of configurations we care about.
+let configurations = [
+  {maxVersion: 4, fallbackLimit: 4, website: "enabled.tls13.com"},
+  {maxVersion: 4, fallbackLimit: 4, website: "disabled.tls13.com"},
+  {maxVersion: 3, fallbackLimit: 3, website: "control.tls12.com"}
+];
+
+// some fields are not available sometimes, so we have to catch the errors and return undefined.
+function getFieldValue(obj, name) {
+  try {
+    return obj[name];
+  } catch (ex) {
+    return undefined;
+  }
+}
+
+// extracting the root certificate from the chain
+function extractRootCert(cert) {
+  let root_cert = cert;
+
+  while (getFieldValue(root_cert, "issuer")) {
```

Comment:
I mentioned over email that nsIX509Cert.issuer isn't the best API to use. nsIX509CertDB.asyncVerifyCertAtTime would be better.